### ISSUE TITLE
Spotlight: add null case business logic to similarReflections

### ISSUE THREAD

**Issue by mattkrick** [ParabolInc/parabol#5183]
Predecessors: #5182

Now that `similarReflections` just returns all reflection groups in the meeting, we need to filter that down if the searchQuery is an empty string.

AC
- [ ] if the searchQuery is an empty string, then we use hierarchical agglomeration to filter & sort the reflectionGroups in the meeting (see https://github.com/ParabolInc/parabol/blob/db45e3b5a4c27e917dcc267d33175bae8a04970b/packages/server/graphql/mutations/helpers/handleCompletedStage.ts#L40 as an example)
- [ ] The heuristic above should attempt to return 10 reflections & if the agglomeration returns less (and there are more than 10 other reflection groups in the meeting), then increase the grouping threshold iteratively until it works
- [ ] on the front end, the groups show up as expected in the results space
- [ ] on another browser, if 2 of those groups are merged in real time, that change is reflected in the search results
- [ ] on another browser, if 1 of those reflections is ungrouped, then that change is reflected in the search results

Estimate: 8 hours

**Comment by nickoferrall** [comment]
I'd be keen to work on this one if there's no objections! Will finish the tasks I'm currently working on and aim to start tomorrow

**Comment by nickoferrall** [comment]
Hey @mattkrick, just want to check whether you're thinking that we'll use the `groupReflections` function here or just borrow some of the hierarchical agglomeration logic?

If we are using the `groupReflections` function, it looks like it sorts reflection groups into an order where groups with similar titles are closer to each other (useful for when going from reflect to group stage). 

However, in the Spotlight, I imagine that we want to sort groups by the ones that are similar to the selected reflection title instead?

**Comment by mattkrick** [comment]
yeah, 2 approaches:
- use the grouping heuristic, find the group that the target reflection is in & keep increasing the coefficient until a satisfactory number of cards are returned
- come up with a new heuristic like a k-nearest-neighbors

the first one should be faster & since this is just a first pass that should be AOK.

**Comment by nickoferrall** [comment]
I think that the `reflectionGroup` function isn't working as expected: 
- I get exactly the same output if I change the `groupingThreshold` from 0.25 to 0.99 or 0.001 or 1.5. 
- the smartTitle and title in the groups are always the content of the first reflection
- the `oldReflectionId` and the `reflectionId` only differ in the first 5 `groupedReflections`. All other groups have a sortOrder of 0 and the same two reflectionIds. The seems to be caused by `const MAX_GROUP_SIZE = 5` in [getGroupMatrix](https://github.com/ParabolInc/parabol/blob/d2010c0232870243dfab4e8ddd99063577d3882b/packages/client/utils/smartGroup/getGroupMatrix.js#L18)
- I can see that similar reflections in the retro only seem to be being grouped when they are in the first 5 reflections. If I type "1", "random", "1" as the first three reflections, "1" and "1" will be positioned next to each other when moving to the Group stage. If they are not in the first 5 reflections, their position doesn't move 

I'm using the function in essentially the same way as `handleCompletedStage`:
```
          const reflections = await r
            .table('RetroReflection')
            .getAll(meetingId, {index: 'meetingId'})
            .filter({isActive: true})
            .orderBy('createdAt')
            .run()
            
           groupReflections(reflections, 0.25)
```

My plan is to:

- familiarise myself further with how `groupReflections`, `computeDistanceMatrix`, and `getGroupMatrix` functions are working
- play around with the `MAX_GROUP_SIZE` and `groupingThreshold` 
- if this becomes a blocker, work on the frontend part using regular `reflectionGroups`

![0 75](https://user-images.githubusercontent.com/39854876/132000966-91c5caa5-44f7-40b8-8b1f-c15b11e940ef.png)
![0 75 second](https://user-images.githubusercontent.com/39854876/132000962-a2b77714-f3b4-4a2e-b973-7ce92e1a739a.png)
![0 25](https://user-images.githubusercontent.com/39854876/132000971-db9183e2-d1fb-4794-acdc-51ee8faa2dee.png)
![reflections](https://user-images.githubusercontent.com/39854876/132001110-74b5cc9e-cafd-47ac-a76d-997f8213b9c3.png)

**Comment by nickoferrall** [comment]
Hey @mattkrick, it looks like `groupReflections` is only working for the first five reflections and the `groupingThreshold` doesn't affect the number of groups returned. I'd be happy to work on a fix but I want to check whether you think the logs in those screenshots look off in case I'm mistaken and spend more time down a rabbit hole? 

I'm currently working on the rest of the issue using the `groups` that `groupReflections` returns. 

One difficulty with dragging and dropping is that the `reflectionGroups` live in both the kanban and the Spotlight. Grouping and ungrouping causes a Relay console error stating that there are two records with the same id.

Approaches that I'm working on to solve this:

1. Groups only exist in one place: when opening/closing the Spotlight, remove/add the `similarReflectionGroups` from the `reflectionGroups` stored in Relay. This way, each group only exists on either the kanban or the Spotlight

2. In the `similarReflectionGroups` query in User, return a new id for each group and add an additional field like `oldReflectionGroupId` to connect the group in the Spotlight with the group in the kanban

3. Look into cloning reflection groups like we do with remote reflections

**Comment by mattkrick** [comment]
yep, can change the limit here: https://github.com/ParabolInc/parabol/blob/d2010c0232870243dfab4e8ddd99063577d3882b/packages/client/utils/smartGroup/getGroupMatrix.js#L18

> Grouping and ungrouping causes a Relay console error stating that there are two records with the same id.

this shouldn't be, as relay doesn't care about how many times a group repeats in the UI. Any code snippet you can point me to?

**Comment by nickoferrall** [comment]
Ah it looks like this is actually an issue in master rather than this PR. If I group or ungroup, I get the error: `RelayResponseNormalizer: Invalid record. The record contains two instances of the same id: ...`. I'll look into this.

I've pushed a WIP PR https://github.com/ParabolInc/parabol/pull/5376 but please don't feel obliged to dig in. I was just sharing my progress.

The main thing I wanted to check with you was whether the console logs from `groupReflections` look suspicious to you? Adjusting the `groupingThreshold` from 0.1 to 0.9 doesn't change the output for me and all of the titles and smartTitles that are in the groups are the same as the first reflection.
